### ISSUE TITLE
State of the beta

### ISSUE THREAD

**Issue by leamas** [Rasbats/shipdriver_pi#306]
Think we need a general discussion forum for the beta. Correct me if I'm wrong.

I have tried to apply the current templates to two "complicated" plugins: radar_pi and o-charts_pi. Both of them means meeting the needs of another developer, Kees and Dave so they are reasonably tough test cases. I have found some bugs, but they are now fixed.

Still worried about circleci. Although they have enabled the MacOS builds, my account is till classified as 'free' rather than the former 'open-source'. This means a low limit on usage -- I will actually soon not be able to do any more builds in November.  We need to keep doors open until this is sorted out. My ticket is ongoing, but with long delays. 

Another thing is that we need to test update-templates together. Believe me, it saves a *lot* of time when updating a plugin. It has a learning curve, but not worse than that Rick now seems happy using it.

Thoughts?

**Comment by Rasbats** [comment]
One thing to be aware of when using _update-templates_ is that the keys in _./ci_ no longer exist after copying from ShipDriver.
Need to ensure keys are copied to ./build-deps before using _update-templates_ for the first time with a plugin repo.

**Comment by leamas** [comment]
Yes. Could you add a note to UPDATE-TEMPLATES? 

Keys are keys, even if encrypted. While you of course copy your own keys, there is no point that my keys are everywhere. So this step probably need to stay manual.

I have made a major rewrite of update-templates which more like copies files, avoiding to merge the history. But I cannot test it. see #308.

**Comment by leamas** [comment]
Seems that we are heading for beta2 after the current work in #318. I still have some work to do on the local builds, though.

**Comment by leamas** [comment]
After #320 and #319 my idea is that we should make tag this as beta2. Thoughts?

**Comment by leamas** [comment]
New state after some merging, discussion in #318 etc.: After merging #321 and #322  we have basically no open issues besides documentation. 

Ergo:  After merging #321 and #322 we should tag master  as sd3.0.0-beta2. Or?

**Comment by Rasbats** [comment]
Later.

**Comment by Rasbats** [comment]
Getting Android to write text on the viewport is becoming a "Holy Grail". No managed Android plugin I have used has achieved this. Grib_pi does it but using part of that have not yet helped.  We really need to have a reference plugin, based on the SD workflow that achieves this. This is why I am holding off on any new beta.

**Comment by leamas** [comment]
> Getting Android to write text on the viewport is becoming a "Holy Grail"

If this is such an important issue there should IMHO be an issue filed about it (perhaps by updating #318).

That said, to me it it definitely seems like an example on feature creep. We rightfully postponed  changes in the library handling to after 3.0. IMHO we should do the same here: See that 3.0.0 gets out of the doors before adding still more functionality like this. 

With 5.6.0 being upon us any time we really need to release this. I don't want to make any PRs to Kees' and Dave's plugins until we have released 3.0 for obvious reasons. OTOH, this Android viewport thing is just not that interesting for them (radar_pi does not build on Android anyway).

What we could do here is to  branch the 3.0 release. That  is,  create a  new branch called  v3.0 here and now. I could take the task to maintain this so it goes to a  final release, adding bugfixes etc. While you continue your work on master branch, heading for 3.1 or so. 

Having a release branch like v3.0 is a standard workflow.

Thoughts?

**Comment by Rasbats** [comment]
Couple of things going on here. I needed an updated version of SD available for 5.5.x. Because of that sd3.0.0 is in the catalog. 

Using the function `DrawText()` from _Grib_pi_ , with a small tweek for `glTexEnvi`, I can now draw text on Android  _oplaydo2_pi_. _pidc.cpp_ and _pidc.h_ are now in the _src_ directory, not _plugingl_. 

No need for DC (pidc) in ShipDriver. It should be the responsibility of an individual plugin to include it, if it is needed. It does not impact on the use of OpenGL with any plugin.

Have now removed _pidc_  from SD, amending plugingl files, including _plugingl/CMakeLists.txt_. The opencpn-libs repo needs updating at some point. I will make a PR for this.

>we should tag master as sd3.0.0-beta2. Or?

Are we now at sd3.1.0-beta1? Tested the alpha on msvc and Android. They look good. Builds are here:
[https://cloudsmith.io/~opencpn/repos/shipdriver-alpha/packages/](https://cloudsmith.io/~opencpn/repos/shipdriver-alpha/packages/
)

**Comment by leamas** [comment]
> Are we now at sd3.1.0-beta1? Tes

No. We have not released 3.0.0, so next stop is sd3.0.0-beta2.

If you are ready to go, let's go. Here are the actions I think you should make now:

  - Update Changelog.md with one or two oneliners on what you have done.
  - Still in ChangeLog.md,  change the TBD beta2 date to today's date (assuming you do this today).
  - Commit everything, make sure the repo is clean.
  - Tag with sd3.0.0.beta2 and push the new tag using `git push --tags`
  
This means that we have released our second beta. From now on we should be in a bug-fix only state, not adding any new functionality. 

We should also be prepared to fork the v3.0 branch at any point, at latest when making the 3.0.0 release.

If required, we can make a minor 3.0.1 release *after* 3.0.0 to catch up possible version mess in the catalog. This is simple once 3.0.0 is out..

**Comment by leamas** [comment]
>  I needed an updated version of SD available for 5.5.x. Because of that sd3.0.0 is in the catalog.

There certainly must have been other options than pushing the 3.0.0 tag for this? You should have been able to use any any version higher  than current to achieve this (?)

Now the 3.0.0 tag exists in the remote repo. You need to remove this asap. One way is `git push origin :/refs/tags/sd3.0.0`, there are other ways. But, please remove it, it really messes things up.

Handling versions and tags is complicated enough anyway without having future tags already in the history..

**Comment by Rasbats** [comment]
sd3.0.0-beta2 is now on github. 
Build packages are here:
[https://cloudsmith.io/~opencpn/repos/shipdriver-beta/packages/](https://cloudsmith.io/~opencpn/repos/shipdriver-beta/packages/)

**Comment by leamas** [comment]
... and the premature sd3.0.0 tag is gone. Looks great. Thanks for cleaning up!

Now, let's scrub remaining bugs. I have some at hand, but no  time now...

**Comment by leamas** [comment]
I have tested on the most demanding plugins: radar, oesenc and o-charts. The plugins do have their problems, but I cannot discover  any bugs in in the templates.

@Rasbats : What about you? Ready for the 3.0.0 release? 

If so, please make a note in the changelog like 

    3.0.0  Nov 26, 2021
    * bugfixes

reset PKG_PRERELEASE to "", commit,  check that repo is clean, tag with sd3.0.0 and  `git push --tags` 

And, finally, close this bug...

**Comment by Rasbats** [comment]
In the configuration file _opencpn.ini_ there is a section _**[Settings/ShipDriver_pi]**_. This means that this section appears alongside the main OpenCPN settings.
To appear like other plugins, and keep in the same area of the ini file, this section should be named _**[PlugIns/ShipDriver_pi]**_ (Capital I) and have the settings listed. 
There is a long history of using Settings for my plugins. Gradually changing to PlugIns. Now doing this before making the sd3.0.0 release.

**Comment by leamas** [comment]
Seems that this is in the undefined zone between the templates and the "real" shipdriver stuff. I assume it will not affect other plugins using the templates (?)

If it does, please file a separate issue so it's possible to refer to it in the changelog etc.

**Comment by leamas** [comment]
Is the Settings thing fixed in 244f20f ? If so, is it then time to make the 3.0.0 release?

**Comment by Rasbats** [comment]
sd3.0.0 release pushed to github. Commit 88d3e47

**Comment by leamas** [comment]
OK. Time to celebrate :). 

Unfortunately, the Changelog didn't make it to the release... shit happens. Well amend in next point release, 3.0.1 or so.

**Comment by leamas** [comment]
Anyway, the state of the beta is Closed, we have released 3.0.0. Closing this bug.
